Add TypeScript types to polyfill

justingrant opened the pull request:
_(re-creating this PR because I accidentally closed #616)_
 
Fixes #581 with TypeScript types!  This PR models the full API of Temporal, but it's a big surface area so I'll need reviewer help to make sure I didn't miss any parameters or methods or get any types wrong.

I found that seeing all the types in one place was useful to reason about the API surface as a whole. I encourage looking through to make sure that the actual API matches what you think the API should be. 

I picked what seemed like appropriate names for options types. Feel free to bikeshed on the names. Figuring out which options types were shared by which methods was a useful exercise to understand the relationships between different methods and types. It really helped me understand the object model better.

My apologies for all the GH issues I filed while building these types. It's hard to go through every type, field, method, and parameter without coming across a lot of questions!  For issues I raised about specific methods or properties, I marked these with `TODO:` comments in the typings file with pointers to the relevant issue number.

I added JSDoc comments for all the options types, because these are IMHO were the most complex parts of the API... and also they are changing much more than methods and fields. Documentation for the actual Temporal types like `DateTime` can come in a follow-up PR later but for now I wanted to get the basic types into the repo so folks can start to review. 

One problem with TS types in a fast-evolving JS library is keeping the API in sync with the types. Once they start drifting apart, it's very labor intensive to get them back in sync. 

Therefore, I'd recommend that all future PRs that change the API surface must also rev index.d.ts. None of the types are any fancy TS wizardry so this should be do-able even for folks who don't use TS.  If that won't work, then perhaps we could have a tag on PRs to call out which PRs change the API surface, which would make it easier to keep index.d.ts in sync?

I'd also suggest that PRs that change behavior of options types should edit the JSDoc for those types, just like the HTML documentation is edited with new PRs.

ryzokuken commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+    disambiguation: 'balanceConstrain' | 'balance';
+  };
+
+  export interface DifferenceOptions<T extends string> {
+    /**
+     * The largest unit to allow in the resulting `Temporal.Duration` object.
+     *
+     * Valid values may include `'years'`, `'months'`, `'days'`, `'hours'`,
+     * `'minutes'`, and `'seconds'`, although some types may throw an exception
+     * if a value is used that would produce an invalid result. For example,
+     * `hours` is not accepted by `Date.prototype.difference()`.
+     *
+     * The default depends on the type being used.
+     *
+     * TODO: document which types use which defaults.
+     */
+    largestUnit: T;
+  }
```
oops, I understand now what it means.

justingrant replied:
Yep, what's especially nice about using string literal union types in cases like this is the IDE experience where you get autocomplete for the string values. For example, `{disambiguation: 'earlier'}` only requires 6 keystrokes, including the brackets. You also get compiler errors if you either make a typo in a literal string value or if you supply a literal string that's not allowed in that context, e.g.` 'day'` for `YearMonth`.  

We don't have any types (yet) where it would help, but one other nice thing about using literals like this is you can encode incompatible options in the type system, e.g. if `{foo: 'x'; bar: 'y'}` is illegal, you can write a type that shows compiler errors if those options are used together.

ryzokuken commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+    readonly daysInYear: number;
+    readonly isLeapYear: boolean;
+    with(temporalYearMonthLike: YearMonthLike, options: AssignmentOptions): Temporal.YearMonth;
+    plus(temporalDurationLike: YearMonthLike, options: ArithmeticOptions): Temporal.YearMonth;
```
```suggestion
    plus(temporalDurationLike: DurationLike, options: ArithmeticOptions): Temporal.YearMonth;
```

justingrant replied:
Shoot, that's a bad one. Nice catch.

ryzokuken commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+  };
+
+  export class YearMonth extends Required<YearMonthLike> {
+    static from(item: string | object, options?: AssignmentOptions): Temporal.YearMonth;
```
`YearMonthLike` too?

justingrant replied:
Yep, will fix... and will think twice in the future about using search-n-replace for anything other than renaming!

ryzokuken commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+  };
+
+  export class MonthDay extends Required<MonthDayLike> {
+    static from(item: Temporal.MonthDay | string | object, options?: AssignmentOptions): Temporal.MonthDay;
```
```suggestion
    static from(item: MonthDayLike | string | object, options?: AssignmentOptions): Temporal.MonthDay;
```

justingrant replied:
Yep. Per earlier discussion, I'll make this `Temporal.MonthDay | MonthDayLike`

justingrant replied:
Circling back, here's where it ended up:
* Parameters of methods like `plus` that accept a plain-JS property bag will get this signature: `Xxx | XxxLike`. This is redundant, because Temporal Types (except TimeZone and Absolute) all implement their Like interfaces. But I think this redundancy will help users who look at the types in IDE tooltip help, because it makes it easier to realize that they don't have to call `getFields()` to call the method but instead can just pass the class instance.
* Exception: `with` methods accept a Temporal class instance, but that seems like an anti-pattern (users who want to clone an object should use `from`), so I made the signature just `XxxLike` to emphasize that the property bag is the main use case for that method. Passing a class instance won't trigger a TS warning (because every instance type implements its Like interface) but may be useful to guide developers towards best practices.

Holler if you disagree.

ryzokuken commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+  };
+
+  export class Time extends Required<TimeLike> {
+    static from(item: Temporal.Time | string | object, options?: AssignmentOptions): Temporal.Time;
```
```suggestion
    static from(item: TimeLike | string | object, options?: AssignmentOptions): Temporal.Time;
```

justingrant replied:
Same here too.

justingrant commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+     *
+     * - In `'constrain'` mode, out-of-range values are clamped to the nearest
+     *   in-range value.
+     * - In `'balance mode'`, out-of-range values are resolved by balancing them
```
Thanks @ryzokuken for reviewing!  I'm new here, so what's your preferred workflow for updating PRs with review changes?  Should I push additional commits without force-pushing until all review comments are resolved, and then do a final force-push?  Force-push each commit? Or something else?

ryzokuken replied:
Just keep the new changes in new fixup commits, I'll squash and merge. Thanks.

justingrant replied:
@ryzokuken - one more process question: what's the right etiquette around resolving conversations?  Should I resolve conversations that I believe are resolved, or should I wait for the reviewer(s) to resolve if you and/or @ptomato are satisfied?  I'm about to check in an update later today which should fix most of the concerns brought up in the review.

ryzokuken replied:
You may feel free to mark a conversation as resolved. If someone disagrees with your opinion, they can just unresolve, I guess.

justingrant commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+    nanoseconds?: number;
+  };
+
+  export class Duration extends Required<DurationLike> {
```
The advantage of using extends is that it ties the DurationLike and Duration types together, so that if a change is made to one that doesn't match changes to the other (e.g. renaming a field) then you'll get a compiler error to catch the problem.

BTW, how to describe classes in TS is something I'm still not 100% sure about. I was going to research it today. One option is doing what's in the current commit: a `class` keyword with all type info inside. The other option is to create 2-3 types (one for the class constructor+statics, one for instance methods and fields, and an optional third `declare var` for global objects like `Array`), which essentially matches the old-school JS way to create classes. The second way of describing types is what TS's own standard types for ES built-in objects use.  So I'm trying to figure out if they only do it that way for backwards compat wtih ancient TS versions, or if there's some advantage of doing it that way. 

Depending on what I learn (see 
https://stackoverflow.com/questions/62006825/pros-cons-of-declaring-class-types-in-d-ts-using-class-foo-vs-separate-foo) it might make sense to change to the 2-to-3 type way before finalizing this PR.

ptomato replied:
(Also a TypeScript noob here) Are you still going to get an error if the fields change, even if they are overridden readonly below?

justingrant replied:
Oh, I didn't really answer your question.  Yes, it probably makes sense to add `Duration | DurationLike` to all places where we accept a `DurationLike` parameter.  In practice it won't change any type-checking behavior do make this change, but I agree with you that it would be clearer for folks who are reading the type info or viewing it in an IDE. 

I'll scrub all the use of `*Like` types and make this change.

justingrant replied:
> @ptomato Are you still going to get an error if the fields change, even if they are overridden readonly below?

`readonly` refers to what the client code is allowed to do. It's what `get` properties look like in TS types. No problem if the underlying value changes--`readonly` just means that userland code isn't allowed to make those changes. If I try to write code that tries to change a readonly field, I'll get an error in the IDE and compiler. Like this:

![image](https://user-images.githubusercontent.com/277214/82836970-9e77b500-9e7c-11ea-8084-20f875ca0109.png)

justingrant replied:
> It it necessary to use the `extends` clause? Can't we manually specify for each function that accepts a `DurationLike` that it may also accept a `Duration`?

One more thing: I just realized I was using `extends` but should have been using `implements`.  The former is used when you're inheriting from a class, while the latter is for when you're implementing an interface. I'll fix this along with the other issues noted above.

ryzokuken replied:
Exactly. That's why I was confused, because you usually `extend` an existing class...

justingrant replied:
> (Also a TypeScript noob here) Are you still going to get an error if the fields change, even if they are overridden readonly below?

@ptomato - to answer your question more directly, here's what you'll see if you try to rename a field in `Duration` without a  corresponding change in the `DurationLike` interface.

![image](https://user-images.githubusercontent.com/277214/82900562-9c136a80-9f11-11ea-86af-61a79a71744d.png)

justingrant replied:
> Exactly. That's why I was confused, because you usually extend an existing class...

Apologies for misunderstanding your original comment.  I thought you were objecting to the dependency between the class vs. the interface, not to the incorrect `extends`. I almost never write classes in my own projects so I'm still adjusting to the quirks of TS when applied to classes. In class-less code, `extends` is the normal way to way to connect types together and `implements` is almost never used.  With classes it's the reverse.

Normally ESLint (which has great TS support now) would have caught this mistake but I discovered yesterday that my global ESLint config was ignoring .d.ts. This is a non-issue for my own projects where local ESLint config is part of the project, but caused false negatives in Temporal whose ESLint config ignores TS.

I'm looking into what it would require to add TS support to Temporal's ESLint config.  It's harder than I'd hoped because Temporal uses the babel-eslint parser which has more limited linting support than @typescript-eslint/parser. But if I can figure it out then I could add this change into this PR. I suspect the cleanest way to do it would be to create a separate TS-friendly eslint config and then to run ESLint twice in `npm run lint`, once with the regular config and again just for the .d.ts. Or to add a new script in package.json and run both of them in CI. Would something like this be welcome?

ptomato replied:
The `babel-eslint` parser is used specifically because the default parser didn't support `import()`. You're welcome to switch it out for a different parser altogether as long as it lints the existing JS files cleanly. If not, then linting the types file with a separate config in `npm run lint` seems like a good way to go.

justingrant replied:
I got @typescript-eslint/parser to work. It required upgrading eslint to 7.1 (because of https://stackoverflow.com/questions/54514989/) and turning off three lint rules that triggered false-positive warning that weren't real problems in the JS codebase.  But otherwise no changes were required. I'll add these changes to the PR.

justingrant commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+
+  export class Duration extends Required<DurationLike> {
+    static from(
+      item: Temporal.Duration | DurationLike | string | object,
```
Yep, `object` looks like a bug here. I think when I originally wrote that declaration I wasn't sure if it worked like TimeZone where objects are coerced to strings as a fall-through case.  Duration and non-TZ types don't work like that, correct?

ptomato replied:
The way `from` works is:
- `Temporal.Duration`: the duration is cloned
- `DurationLike`: a new `Temporal.Duration` is constructed using the property values
- `string`: parsed as an ISO string
- anything else: is converted to a string first

For the record, `{}` and `{centuries: 5}` both work, and both give a duration of 0. But maybe this isn't the TypeScript way? (As I mentioned, I'm not much of a TypeScript programmer)

What's the best practice for expressing liberally-accepting JS APIs in TypeScript? Should the type be `any` (reflecting the reality of what is possible to pass to the function) or `Temporal.Duration | DurationLike | string` (being explicit about what cases the function handles)?

justingrant replied:
> * `DurationLike`: a new `Temporal.Duration` is constructed using the property values
> * `string`: parsed as an ISO string
> * anything else: is converted to a string first

@ptomato - what determines "duration-like"?  If an object has no Duration properties, but it has a `toString()` implementation, will Temporal try to call it?  If this is "no", then I'll remove `object` from the accepted types here because `Duration` or `DurationLike` are the only `object` types that should be allowed. 

> For the record, `{}` and `{centuries: 5}` both work, and both give a duration of 0. But maybe this isn't the TypeScript way? (As I mentioned, I'm not much of a TypeScript programmer)

`{}` seems reasonable for an empty duration and it's currently allowed in the TS types because all fields are optional for the `*Like` interfaces.

But if you do specify an object literal like `{centuries: 5}` then TS will complain, but only for object literals. If you want to opt into this behavior, e.g. to create a custom DurationLike type, that's fine. Users can also use a type cast, which means "Hey TS I know what I'm doing, stop complaining"/ Here's an example of all three cases:
 
![image](https://user-images.githubusercontent.com/277214/82839273-b1da4e80-9e83-11ea-859a-024e39916a5e.png)

BTW, a big reason to do this is typo protection. For example, this is also how we get IDE or compiler errors for `{year: 5}` (no "s").

Hey, will calendars expose custom fields, like Japanese `era` on `Date`?  If yes, then the way we'd probably solve this in TS is to make types generic, e.g. `DateTimeLike<CalendarType = Temporal.IsoCalendar>` or something like that. This would enables calendars to expose extra fields to the calendar types while still retaining type safety, and would let users assert that they're expecting dates from a particular calendar to have particular fields.

> What's the best practice for expressing liberally-accepting JS APIs in TypeScript? Should the type be `any` (reflecting the reality of what is possible to pass to the function) or `Temporal.Duration | DurationLike | string` (being explicit about what cases the function handles)?

The latter. The rule of thumb is that the types should match the documentation, not the implementation. When you build types, it's essentially like documentation encoded into type form--hints from the library about what developers *should* do.  Just because something doesn't throw an exception doesn't mean it should be encouraged in the TS types.

ptomato replied:
>  If an object has no Duration properties, but it has a `toString()` implementation, will Temporal try to call it?

`Temporal.Duration.from({ toString() { return 'P2D'; }}).toString() === 'PT0S'`, so I think the answer is "no."

justingrant replied:
OK I'll remove `object` in those cases.

ptomato commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+     * The default is `'constrain'`.
+     */
+    disambiguation: 'constrain' | 'balance' | 'reject';
+    // TODO: should change name from `disambiguation` to `overflow`? See #607.
```
I'd suggest removing the TODO notes and instead just have the types file always reflect the current state of the polyfill. (it would be nice if we could fail CI when they get out of sync, but I'm not sure that's possible?)

justingrant replied:
When is the polyfill going out to npm?  It definitely makes sense to remove these comments before publishing to npm.

In the meantime I was assuming that there's value in highlighting which APIs might be changing-- especially those that seem to have initial consensus but which others might object to. I wanted to provide a way to encourage feedback before changes are finalized as opposed to after. Also to prevent duplicate issues from being filed if changes are already being discussed. But I'm also happy to take them out, let me know what you think.

> (it would be nice if we could fail CI when they get out of sync, but I'm not sure that's possible?)

Yep, this only works if there's other code to get out of sync with.  Because the polyfill implementation and tests are untyped JS, it's harder. The TS compiler can type-check plain JS code and find some problems but it's often hit-or-miss.

One way to solve this would be to be able to run at least some test code through the TS compiler-- essentially to be able to write tests in TS or JS. With at least one TS test for every API schema change, we'd probably be able to catch most out-of-date types.

One thing I could also try would be running some of the tests through the TypeScript compiler-- not to actually build tests in TS, but to trigger errors if test code is doing things that TS can flag as errors. 

The CI part is pretty easy: just include a single run of the TS compiler as part of CI build/test scripts. Having the actual content to compile is the hard part. 

One no-brainer thing to do would be to run the .d.ts file itself through the TS compiler. This would at least catch syntax errors or inconsistencies inside the type definitions themselves, which is better than nothing.

Want me to open an issue for this?

ptomato replied:
Sounds like some great ideas. I think it'd be useful to first add validation of the .d.ts file itself, then we can see how difficult it would be to compile the test files with the TS compiler.

justingrant replied:
@ptomato - circling back on this issue... still looking for your reply:

> When is the polyfill going out to npm? It definitely makes sense to remove these comments before publishing to npm.

> In the meantime I was assuming that there's value in highlighting which APIs might be changing-- especially those that seem to have initial consensus but which others might object to. I wanted to provide a way to encourage feedback before changes are finalized as opposed to after. Also to prevent duplicate issues from being filed if changes are already being discussed. But I'm also happy to take them out, let me know what you think.

ptomato replied:
Oh, sorry about that! Yes, we intended to put the polyfill on npm as soon as calendars and Temporal.parse have landed (basically everything in the https://github.com/tc39/proposal-temporal/milestone/4 milestone). Ideally before the TC39 meeting next week.

Personally I'd rather have no TODOs in the types file but instead make it clear that we encourage feedback on all of the API, not just specific parts, and that changes will only be finalized when the proposal advances to Stage 3.

ptomato commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+     * In `'balance'` mode, all fields are balanced with the next highest field,
+     * no matter if they are negative or not.
+     *
+     * TODO: add an example of each mode.
```
Do these doc comments pop up in the editor if you're using a TypeScript-aware editor? Otherwise we can probably just refer to the JS docs for examples.

justingrant replied:
Yep, that's the main value of doing them in JSDoc. You get the docs inline in the IDE which can be helpful. Here's what it looks like: 

![image](https://user-images.githubusercontent.com/277214/82836658-cd415b80-9e7b-11ea-8347-0bb91a1debbb.png)

justingrant commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+      options?: AssignmentOptions
+    ): Temporal.Duration;
+    constructor(
+      years?: number,
+      months?: number,
+      days?: number,
+      hours?: number,
+      minutes?: number,
+      seconds?: number,
+      milliseconds?: number,
+      microseconds?: number,
+      nanoseconds?: number
```
Yep, but from TS's perspective default values are part of the implementation, not the type. It makes sense when you think about it: TS is really just a way to create custom lint rules for each JS type. It just tells you what kind of code is legal vs. illegal. So it can tell you that a parameter is required or not, and what kinds of values are allowed or disallowed. But a default parameter value isn't something that would generate a compiler error so TS ignores it. At least that's how I believe it works. 

I actually made the same mistake when I wrote the first draft of the types and had to Google it to find out why `=0` wasn't compiling. ;-) https://stackoverflow.com/questions/35074365/typescript-interface-default-values

justingrant commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+    toString(): string;
+  }
+
+  export class Absolute extends Required<AbsoluteLike> {
```
Oops, that's a regex search-and-replace gone awry. I introduced this when I added `extends` to the class types. I remembered that TimeZone was special and didn't have a `*Like` interface, but I forgot that Absolute didn't either. I'll remove.

ptomato replied:
There is no such thing as an AbsoluteLike, though, so I think it should just be removed instead :smile:

justingrant replied:
> I think it should just be removed instead
Yep, will remove. This was caused by a search-n-replace gone awry.

justingrant commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+    static fromEpochMicroseconds(epochMicroseconds: bigint): Temporal.Absolute;
+    static fromEpochNanoseconds(epochNanoseconds: bigint): Temporal.Absolute;
+    static from(item: Temporal.Absolute | string | object): Temporal.Absolute;
+    static compare(one: Temporal.Absolute, two: Temporal.Absolute): 1 | -1 | 0;
```
Good catch, I'll fix.

justingrant commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+      options?: DifferenceOptions<'days' | 'hours' | 'minutes' | 'seconds'>
+    ): Temporal.Duration;
+    // Should Absolute and DateTime have different names for `inTimeZone`? See #574.
+    inTimeZone(temporalTimeZoneLike?: TimeZoneLike): Temporal.DateTime;
```
Nope, I agree. How about `tzLike` which will still emphasize that it's OK to pass a string without being too long? 

BTW, I originally got this name from here: https://github.com/justingrant/proposal-temporal/blob/d5f480ae206207fd7561e560c860c348f61ce384/polyfill/lib/now.mjs#L16-L27

In general I picked parameter names by trolling through the Temporal source and seeing if there was a consistent name used for each concept. I didn't necessarily go back and clean up names for public consumption, so if you see others that look amiss, let me know and I'll fix.

ptomato commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+    toString(temporalTimeZoneLike?: TimeZoneLike): string;
+  }
+
+  export type DateLike = {
```
Not necessarily within scope for this pull request, but for calendar-aware types (Date, DateTime, YearMonth, MonthDay) this is about to get a bit more complicated. A few calendars are going to require extra fields (such as `era` for the Japanese imperial calendar) and those will be passed to `from() / with()` and returned from `getFields()`.

justingrant replied:
Haha, I asked the same question in my comment above! I'll delete it there and move that discussion here. 

> Hey, will calendars expose custom fields, like Japanese `era` on `Date`?  If yes, then the way we'd probably solve this in TS is to make types generic, e.g. `DateTimeLike<CalendarType = Temporal.IsoCalendar>` or something like that. This would enables calendars to expose extra fields to the calendar types while still retaining type safety, and would let users assert that they're expecting dates from a particular calendar to have particular fields.

Anyway, where will the extra fields be? On which types?  And what about the Calendar type itself-- where is it documented?

ptomato replied:
Oops, I must have missed that comment! Sorry!

The extra fields can potentially be on Date, DateTime, YearMonth, and MonthDay. The Calendar type (draft API) is currently documented in https://tc39.es/proposal-temporal/docs/calendar-draft.html and there's also ongoing implementation work in progress at https://github.com/tc39/proposal-temporal/commits/calendars

justingrant replied:
OK. Do you think I should add Calendar to TS types now, or wait for the Calendar API to land and make changes then?

ptomato replied:
I think it would be better to wait for the Calendar API to land. (Basically as soon as the TS types land, we should require that any pull requests that make subsequent API changes also update the types file.)

ptomato commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+  export class Date extends Required<DateLike> {
+    static from(item: Temporal.Date | string | object, options?: AssignmentOptions): Temporal.Date;
+    static compare(one: Temporal.Date, two: Temporal.Date): ComparisonResult;
+    constructor(year: number, month: number, day: number);
```
We are calling these constructor arguments `isoYear`, `isoMonth`, and `isoDay` in the spec to emphasize that they are not calendar-aware. They haven't been renamed in the polyfill yet, but probably best to put the correct names in the types file to start with.

justingrant replied:
One concern I had for that naming was that "ISO" for many JS devs means "_that weird string format that ends in Z and has the wrong time that I don't really understand but I know it's used to store in databases or send to external APIs_". So just using "iso" alone will probably confuse more than it will help. My suggestion would be to leave as "year" in the parameter name and then to highlight in the docs that it's an ISO Calendar year. Per @littledan's [notes](https://github.com/tc39/proposal-temporal/issues/268#issuecomment-570670212) about who's actually going to be developing with non-ISO calendars, it seems clear that (unlike, for example, Japanese language localization where the developer may only know Japanese) most developers using non-ISO calendars are also using ISO calendars, so they'll probably already be on the lookout for docs like this because they're already familiar with the challenges of dual-calendar implementations. 

What do you think?

If you do want to change the parameter names, then I'd at least suggest `isoCalendarXXX` to clarify that ISO in this context refers to the ISO calendar and not the ISO string format.

justingrant replied:
@ptomato - I wanted to circle back after our discussion in the other issue. First, should these parameters be renamed in the TS types?  If yes: 
* should they be 'isoXxx' or 'isoCalendarXxx'?  I think the latter might be a little clearer but don't feel very strongly one way or the other. 
* which types' constructor YMD parameters should be renamed? All of them or just a subset?

ptomato replied:
I think `isoCalendarXxx` is not necessary for just the constructor arguments, since we are considering the constructors to be low-level API. And all of them should be renamed, since those types will all support other calendars. So my suggestion would be:
- Date - `constructor(isoYear: number, isoMonth: number, isoDay: number)`
- DateTime - `constructor(isoYear: number, isoMonth: number, isoDay: number, hour?: number, minute?: number, second?: number, millisecond?: number, microsecond?: number, nanosecond?: number)`
- YearMonth - `constructor(isoYear: number, isoMonth: number)`
- MonthDay - `constructor(isoMonth: number, isoDay: number)`

justingrant replied:
OK sound good. Making changes now. I'm assuming that `YearMonth.prototype.withDay` and `MonthDay.prototype.withYear` can stay as plain `day` and `year`, right?

ptomato replied:
That's right, if the calendar is not ISO, then they take calendar days/years, not ISO days/years.

ptomato commented on polyfill/index.d.ts:
```diff
@@ -0,0 +1,385 @@
+
+  export type TimeZoneLike = Temporal.TimeZone | string;
+
+  export class TimeZone {
```
Heads up, I just merged `getPossibleAbsolutesFor(dateTime: Temporal.DateTime): Temporal.Absolute[];` into the main branch.

justingrant replied:
Cool!  I'll update the types.

justingrant replied:
Also, I noticed that `equals` was just committed. I'll add these methods into the TS types.

justingrant commented:
OK, I think I've addressed all the code review comments, modulo two issues I'll need @ptomato's feedback on. Also added ESLint support for the TS types. 

**Types Updates**

Resolve issues found by @ryzokuken and @ptomato reviews
and update types with newly-landed API changes.

* FIx minor JSDoc typos
* Change `extends` -> `implements` in class types
* Remove `object` type from `from` parameters
* Use `Xxx | XXXLike` for params that take a property bag...
* ...except `with` where passing class instances is an anti-pattern
* Remove mistaken `AbsoluteLike` (it doesn't exist)
* Shorten overly-long parameter names
* Replace TimeZoneLike type with just `TimeZone | string`
* Add `compare` methods
* Add `TimeZone.prototype.getPossibleAbsolutesFor` method
* add `export` for all the functions in `now` (fixes lint errors)

**Add linting suport for TS types**

This switches from babel ESLint parser to '@typescript-eslint/parser'
so that index.d.ts can be linted. The same lint rules are active that
were used before, but there are now new TS-specific rules too.
I disabled three TS rules that triggered false-negative issues in
the polyfill JS code, but otherwise uses the default "recommended"
TS rules are used.

Added three new dev dependencies for the parent (not polyfill)
package.json: typescript, the TS eslint plugin, and its parser.

Ran into a bug that was caused by an outdated version of ESLint, so
upgraded ESLint from 6.8.0 to 7.1.0.

justingrant commented:
Fixed a mistake from last night's commit: unlike `compare`, the `equals` methods are instance methods, not static. Ignore the first commit message from this morning-- it had a typo and I amended.

justingrant commented:
OK, just pushed an update which should resolve the two issues remaining from @ptomato: 
* Adds `iso` prefix to YMD constructor parameter names
* Removes TODO comments

Also, I added short JSDoc text (copied from online docs) for each Temporal class type and `now`, each of them linking to the online docs. It's a nice IDE experience when folks are trying to understand the API for the first time, e.g. 

![image](https://user-images.githubusercontent.com/277214/83113165-1a8c1b80-a07c-11ea-9a6c-0bbe5976945e.png)

justingrant commented:
Travis is stuck and isn't reporting status back to GitHub. Anyone know how to fix?

ptomato commented:
I just resolved the merge conflict, which started a new build — hopefully that unsticks it.
